### PR TITLE
AP_Motors: quadratic programming optimal mixer, code optimised

### PR DESCRIPTION
**Pull request by IamPete1**
The key difference between this and the new mixer is how the constraints are applied. The current mixer constrains all motors by the maximum or minimum output. This new approach constrains each motor individually.

This is a demo and explanation, that turned out longer than I planned:
https://youtu.be/oaU7gt31oZk

The new approach should be better with high numbers of motors and for motor failure. It also allows new possibility such as imposing slew limiting at a mixer level. It also makes it much easier to trade priorities between roll, pitch, yaw and throttle.

See https://github.com/ArduPilot/ardupilot/pull/20836

This now runs reasonably fast, this was mostly achieved by doing as much stuff with arrays rather than neat matrices. Octo adds about 10% load on H7. Runs in the order of 200us.

This is only enabled on H7, we could probably get away with F7. I suspect not F4, but we could try...

This adds a flight option to enable and has full rate logging of inputs and outputs along with the run time and number of iterations. We can remove the logging, or make it optional in the future. This now sets the limit flags, although we might want to revisit the thresholds, currently it sets the flag if the achieved output more than 5% away from the request.

I think the key thing now is just to get more testing IRL. Particularly on vehicles with lots of motors where it should be more beneficial and take longer to run. 

This mixer is better than the current one for motor failure, but it could be even better again if we tell it which motors have failed by updating there constraints. But I think that can weight for a future enhancement.

### CODE REVIEW

File: libraries/AP_Motors/AP_MotorsMatrix_Optimal.cpp [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+    limit.pitch = fabsf(inputs[1] - outputs[1]) > threshold;
+    limit.yaw = fabsf(inputs[2] - outputs[2]) > threshold;
+    limit.throttle_lower = (outputs[3] - inputs[3]) > threshold;
+    limit.throttle_upper = (inputs[3] - outputs[3]) > threshold;
+
+    // copy to motor outputs
+    for (uint8_t i = 0; i < num_motors; i++) {
+        _thrust_rpyt_out[i] = x[i];
+    }
+
```
[the remark this answers is not in the document]
Comment (IamPete1) [reply libraries/AP_Motors/AP_MotorsMatrix_Optimal.cpp:305]: done

File: ArduCopter/system.cpp [hunk 2]
```diff
@@ -380,8 +380,16 @@ void Copter::allocate_motors(void)
         case AP_Motors::MOTOR_FRAME_DECA:
         case AP_Motors::MOTOR_FRAME_SCRIPTING_MATRIX:
         default:
-            motors = new AP_MotorsMatrix(copter.scheduler.get_loop_rate_hz());
-            motors_var_info = AP_MotorsMatrix::var_info;
+#if defined(STM32H7) || CONFIG_HAL_BOARD != HAL_BOARD_CHIBIOS
```
Comment (peterbarker) [inline ArduCopter/system.cpp:383]: `#if AP_MOTOR_FRAME_OPTIMAL_ENABLED` or similar...

File: libraries/AP_Motors/AP_MotorsMatrix_Optimal.cpp [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+            }
+            break;
+        }
+        default:
+            return;
+    }
+
+    // update frame name
+    if (frame_string != nullptr) {
+        free(frame_string);
```
Comment (tridge) [inline libraries/AP_Motors/AP_MotorsMatrix_Optimal.cpp:192]: delete []

File: ArduCopter/system.cpp (hunk 2 above)
Comment (peterbarker) [reply ArduCopter/system.cpp:383]: ... also, why aren't we compiling this on SITL?

Comment (tridge) [inline ArduCopter/system.cpp:383]: should be #if AP_WITH_OPTIMAL_MIXER

File: libraries/AP_Motors/AP_MotorsMatrix_Optimal.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include "AP_MotorsMatrix.h"
+#include <AP_Math/matrix_vec_dsp.h>
+
```
Comment (tridge) [inline libraries/AP_Motors/AP_MotorsMatrix_Optimal.h:5]: add #define AP_WITH_OPTIMAL_MIXER 1/0
based on flash size, H7, SITL

**Review by tridge — CHANGES_REQUESTED** [review]
needs testing first with SITL on hw to measure CPU cost, and with SITL valgrind with motors out

**Comment by tridge** [comment]
also test with --enable-malloc-guard on SITL on hw

File: libraries/AP_Math/matrix_vec_dsp.cpp [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    arm_mult_f32(A,B,dest,size);
+#else
+    for (uint16_t i = 0; i < size; i++) {
+        dest[i] =  A[i] * B[i];
+    }
+#endif
+}
+
+// dot product
+void dot_prod(const float *A, const float *B, uint8_t size, float *dest)
```
Comment (tridge) [inline libraries/AP_Math/matrix_vec_dsp.cpp:140]: would be better as a reference

File: libraries/AP_Math/matrix_vec_dsp.cpp [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+// element wise invert
+void vec_inv(const float *A, float *dest, uint8_t size)
+{
+    for (uint16_t i = 0; i < size; i++) {
+        dest[i] =  1.0 / A[i];
+    }
+}
+
+// matrix multiplied by vector
+void mat_vec_mult(const Matrix *A, const float *B, float *dest)
```
Comment (tridge) [inline libraries/AP_Math/matrix_vec_dsp.cpp:173]: reference

File: libraries/AP_Math/matrix_vec_dsp.cpp [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        for (uint8_t j = 0; j < A->numCols; j++) {
+            dest[i] += A->pData[i*A->numCols + j] * B[j];
+        }
+    }
+}
+
+// in place Cholesky factorisation
+// Note that this method does not zero the upper triangle
+// No need with forward_sub and backward_sub implementation
+void cholesky(Matrix *A)
```
Comment (tridge) [inline libraries/AP_Math/matrix_vec_dsp.cpp:186]: reference to matrix

File: libraries/AP_Math/matrix_vec_dsp.cpp [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+            for (uint8_t j = k; j < A->numRows; j++) {
+                A->pData[j*A->numCols + k] -= A->pData[j*A->numCols + i]*A->pData[k*A->numCols + i];
+            }
+        }
+    }
+}
+
+// forward substitution
+// equivelent to: x = a \ b if a is lower triangular
+void forward_sub(const Matrix *A, const float *B, float *dest) {
```
Comment (tridge) [inline libraries/AP_Math/matrix_vec_dsp.cpp:204]: reference

File: libraries/AP_Math/matrix_vec_dsp.cpp [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+        for (uint8_t j = 0; j < i; j++) {
+           dest[i] -= A->pData[i*A->numCols + j] * dest[j];
+        }
+        dest[i] /=  A->pData[i*A->numCols + i];
+    }
+}
+
+// backwards substitution, transposing a
+// equivelent to: x = a' \ b if a is lower triangular
+void backward_sub_t(const Matrix *A, const float *B, float *dest) {
```
Comment (tridge) [inline libraries/AP_Math/matrix_vec_dsp.cpp:216]: reference

**Comment by tridge** [comment]
would be great for quadplanes

File: ArduCopter/system.cpp [hunk 10]
```diff
@@ -380,8 +380,16 @@ void Copter::allocate_motors(void)
...
+            if ((copter.g2.flight_options & uint32_t(FlightOptions::USE_OPTIMAL_MIXER)) != 0) {
+                motors = new AP_MotorsMatrix_Optimal(copter.scheduler.get_loop_rate_hz());
```
Comment (tridge) [inline ArduCopter/system.cpp:385]: might be worth using fast memory, with malloc_type()

**Comment by tridge** [comment]
@IamPete1 @lthall another worthwhile test in RF9 is flying a 4 motor quadplane configured as an octo. More companies are flying with octo setups on quadplanes now, and setup the power so the top and bottom layers are on indepdendent power, and also independent CAN bus for ESC outputs. The idea is to be able to handle a complete layer outage.
Easiest way to test this is to fly a 4 motor quadplane with Q_FRAME_CLASS set to octo, and see if it performs better/worse/no-change in significant wind for takeoff/landing/transition

File: ArduCopter/system.cpp (hunk 10 above)
Comment (IamPete1) [reply ArduCopter/system.cpp]: done, marginally faster

File: libraries/AP_Motors/AP_MotorsMatrix_Optimal.h (hunk 4 above)
Comment (IamPete1) [reply libraries/AP_Motors/AP_MotorsMatrix_Optimal.h:14]: done

File: ArduCopter/system.cpp (hunk 2 above)
Comment (IamPete1) [reply ArduCopter/system.cpp]: done

File: libraries/AP_Motors/AP_MotorsMatrix_Optimal.cpp (hunk 3 above)
Comment (IamPete1) [reply libraries/AP_Motors/AP_MotorsMatrix_Optimal.cpp]: done

**Comment by IamPete1** [comment]
Run the new test under valgrind, no issues.

**Comment by IamPete1** [comment]
Tested a X8, all bottom motors failure. It is worse than the current mixer because it does not use the top and bottom motors symmetrically, So when the top motors are gone it leaves a asymmetry. This would be fixed if we told it which motors are failed.

**Comment by IamPete1** [comment]
For comparison this is the two mixes on a octo quad if we tell the new mixer it has lost the bottom motors. 

image

The original roll/pitch gradient is maintained as much as possible. The current mixer half's the gradient because it does not compensate for the lost motors. So it is effectively halving your PIDs. So, new mixer should behave closer to the no motor loss vehicle. 

Of course, to take advantage of this we need to workout a way to tell the mixer its lost motors.....

**Comment by IamPete1** [comment]
A sim on hardware log, only a Quad, but seemed to run fine.  Solve took on average 66us:

image

Three or four iterations:
image

Always converged:
image


[Sim on Hardware Quad +.zip](https://github.com/ArduPilot/ardupilot/files/8949167/Sim.on.Hardware.Quad.%2B.zip)

**Comment by tridge** [comment]
@IamPete1 these logs from Michelle include using flip mode, we hoped that would exercise things nicely

**Comment by IamPete1** [comment]
@MichelleRos @tridge Thanks for testing! Looks like there were no numerical problems, only one case where it failed to reach the stop tolerance and did the full 10 iterations, it must have been very close. No long loops and running in 50 to 100 us. 

@MichelleRos Do you have a feel for if it was flying better or worse? I would not expect it to be any better on a quad, but at least it should not be worse.

**Comment by IamPete1** [comment]
A little update, I have a version of this that runs using ECOS. https://github.com/embotech/ecos it is a conic solver, allowing us to extend this methodology to tilting motors. I have a MATLAB tricopter mixer using ECOS. Branch that uses it in place of the DIY solver here: https://github.com/IamPete1/ardupilot/tree/matrix_solve_ecos. However, its too slow. Approx 100 times slower than the solver in this PR. 

The point of moving to ECOS was to try and avoid understanding the maths, but I think to get conic solving working we will have to understand the maths sufficiently to strip out a bunch of ECOS to make it fast, at which point we might be better extending the solver in this PR.

ECOS was my best find for a light weight solver that could be made fast, but it really is still too complex, it is setup for problems with hundreds or thousands of variables rather than the ten or so we have here. If anyone had any other suggestion I would love to hear them.

Basically I don't think were going to improve on the speed of the solver in this PR significantly. However it would have been nice if we could use the same solver for this mixer as for a tilting motors mixer. However it seems that that will be significantly more effort again, and possibly come with a speed disadvantage.

So, from a code maturity and testing point of view I'm happy with this code. I think it should work as advertised. There are some enhancements that it would benefit from, such as motor failure detection and in-mixer slew limiting.

However, I would really like to get some testing that shows this mixer is a noticeable improvement for some vehicles in the real world before we merge.
